### PR TITLE
Add value observer extensions.

### PR DESCRIPTION
**Pull request by iwiwi**
This PR adds new extensions: `observe_value`, `observe_time`, and `observe_lr`. Extensions `observe_time` and `observe_lr` record the elapsed time and current learning rate to trainer's `observation`. For example, together with `PrintReport`, the elapsed time and current learning rate are displayed during experiments like the following figure. I personally always use these extensions for my experiments.

image

Extension `observe_value` is a more general one which receives a function that returns a value to record.

### CODE REVIEW

File: chainer/training/extensions/value_observation.py [hunk 1]
```diff
@@ -0,0 +1,45 @@
+from chainer.training import extension
+import time
```
Comment (unnonouno) [inline chainer/training/extensions/value_observation.py]: Split system import and project import.

```
import time

from chainer.training import extension
```

Comment (iwiwi) [inline chainer/training/extensions/value_observation.py]: Thanks! updated!

File: chainer/training/extensions/value_observation.py [hunk 2]
```diff
@@ -0,0 +1,46 @@
...
+    """
+    start_time = time.time()
+    return observe_value(key, lambda _: time.time() - start_time)
+
+
+def observe_lr(optimizer, key='lr'):
+    """Returns a trainer extension to record the learning rate.
+
+    Args:
+        optimizer: Optimizer object whose learning rate is recorded.
```
Comment (unnonouno) [inline chainer/training/extensions/value_observation.py]: Add link to the manual `optimizer (~chainer.Optimizer)`

File: chainer/training/extensions/value_observation.py [hunk 3]
```diff
@@ -0,0 +1,46 @@
...
+from chainer.training import extension
+
+
+def observe_value(key, target_func):
+    """Returns a trainer extension to continuously record a value.
+
+    Args:
+        key (str): Key of observation to record.
+        target_func (function): Function that returns the value to record.
+            It must take one argument: trainer object.
```
Comment (unnonouno) [inline chainer/training/extensions/value_observation.py]: Use `:class:~chainer.Trainer`

File: chainer/training/extensions/__init__.py [hunk 4]
```diff
@@ -6,13 +6,17 @@
...
 from chainer.training.extensions import progress_bar
+from chainer.training.extensions import value_observation
 
 
 dump_graph = computational_graph.dump_graph
 Evaluator = evaluator.Evaluator
 ExponentialShift = exponential_shift.ExponentialShift
 LinearShift = linear_shift.LinearShift
 LogReport = log_report.LogReport
+observe_value = value_observation.observe_value
```
Comment (unnonouno) [inline chainer/training/extensions/__init__.py]: How about sorting with right hand side value? I mean, write `value_observation.xxx` after `progress_bar.xxx`.

**Comment by unnonouno** [comment]
Is it possible to write tests for these extensions?

**Comment by iwiwi** [comment]
It is not impossible to write tests, but, given the fact that currently other reporting extensions (e.g., LogReport) do not have tests, I think it is out of the scope of this PR. https://github.com/pfnet/chainer/tree/master/tests/chainer_tests/training_tests/extensions_tests

I fixed other points. Thanks for the comments!

File: chainer/training/extensions/value_observation.py [hunk 5]
```diff
@@ -0,0 +1,46 @@
...
+        key (str): Key of observation to record.
+        target_func (function): Function that returns the value to record.
+            It must take one argument: :class:~chainer.training.Trainer object.
+    Returns:
+        The extension function.
+    """
+    @extension.make_extension(
+        trigger=(1, 'epoch'), priority=extension.PRIORITY_WRITER)
+    def _observe_value(trainer):
+        trainer.observation[key] = target_func(trainer)
```
Comment (beam2d) [inline chainer/training/extensions/value_observation.py]: Can we support a case that `target_func` returns multiple values (possibly with additional keys)? It can be a different extension.

File: chainer/training/extensions/value_observation.py [hunk 6]
```diff
@@ -0,0 +1,46 @@
...
+        key (str): Key of observation to record.
+
+    Returns:
+        The extension function.
+    """
+    start_time = time.time()
+    return observe_value(key, lambda _: time.time() - start_time)
+
+
+def observe_lr(optimizer, key='lr'):
```
Comment (beam2d) [inline chainer/training/extensions/value_observation.py]: Optimizer can be extracted via trainer: `trainer.updater.get_optimizer('main')`. When user sets multiple optimizers to the updater, we should specify the key other than `'main'`. So how about accepting the optimizer name instead of the optimizer object itself, and making `'main'` the default?

File: chainer/training/extensions/value_observation.py (hunk 5 above)
Comment (beam2d) [inline chainer/training/extensions/value_observation.py]: (this can be deferred to another PR)

Comment (iwiwi) [inline chainer/training/extensions/value_observation.py]: Thanks for the feedback! May I ask what do you mean concretely by 'multiple values'. I suppose it is not a tuple, so is it dict of key and value?

File: chainer/training/extensions/value_observation.py (hunk 6 above)
Comment (iwiwi) [inline chainer/training/extensions/value_observation.py]: Updated!

**Comment by iwiwi** [comment]
I updated the code to reflect one of @beam2d -san's comment. The other is currently not clear to me, in particular, what kind of return value types should be considered.

**Comment by beam2d** [comment]
I meant there would be a case that one wants to set multiple values to the observer by one extension (e.g. report both mean and variance of a set of some values). I might misunderstand the purpose of this extension, though. With the current design, we have to write an extension from scratch when we want to report multiple values at once, but it might be ok.

**Comment by iwiwi** [comment]
I'm sorry but I still do not fully understand what you are saying. What do you mean by 'one extension'. For example, more than one value can be monitored by this `observe_value` extension as follows:

> trainer.extend(extensions.observe_value('hoge', lambda _: 10), trigger=log_interval)
> trainer.extend(extensions.observe_value('piyo', lambda _: 20), trigger=log_interval)

**Comment by unnonouno** [comment]
I tested this PR on my local environment with imagenet example, and it worked well. As @iwiwi -san says, we don't have tests for extensions. We need to make a test mechanism for trainer-related classes, but it is another issue.
I think this PR looks good. @beam2d -san, do you have any other concern?

**Comment by beam2d** [comment]
I discussed with @iwiw offline and decided to add the timing functionality to LogReport instead of an independent extension. I also discussed on supporting multiple value observations, and agreed on separating it into another extension (maybe named observe_values) and defer it to another issue.

**Comment by beam2d** [comment]
LGTM now!

**Comment by iwiwi** [comment]
Thank you @unnonouno -san and @beam2d -san for reviewing and testing!
